### PR TITLE
Change how request content is read in libcurl

### PR DESCRIPTION
**Pull request by stephentoub**
libcurl wants to be able to call back to CurlHandler at arbitrary times to ask it to seek in the request content stream.  This can happen for redirects, for auth challenges, etc.  Handling this has been a challenge due to the way HttpContent is designed.

HttpContent provides two modes for accessing the data via Streams:
- Push: You call CopyToAsync(Stream, TransportContext) to have the data from the HttpContent copied into the target stream.
- Pull: You call ReadAsStreamAsync to get a stream that represents the data in the HttpContent, and you can read from that.

The pull model is ideal for libcurl.  It allows us to get a stream which we can jump around in if it's seekable, and we can request a new stream if it's not seekable. The push model is challenging, as we're forced to put in plcae an intermediary, a push-pull adapter that let's us effectively read from a type that's trying to write to us.

There is however an issue with the pull model.  With the push model, StreamContent's CopyToAsync tracks whether the stream has already been accessed, and if it has and if it's non-seekable, trying to call CopyToAsync again throws to alert you that rewinding simply isn't possible.  The pull StreamContent's ReadAsStreamAsync, however, does not.  This means that if you get a non-seekable stream and you call ReadAsStreamAsync again, you get back the same stream (or, rather, a read-only wrapper around the same stream) that is at the existing position rather than the original position.  Then resending that content will end up sending less data than it should have, resulting in failure, just not abruptly.

That issue caused us to go down the path of the more challenging push model and trying to use CopyToAsync.  However, recently that's been shown to have even worse effects.  We don't have the ability to cancel CopyToAsync, which means that if the first copy is still in-flight when we try to seek, we have no choice but to fail, even if the stream is actually seekable, as most are (e.g. memory streams, file streams, etc.)  This can happen deterministically in cases where libcurl decides to redirect before it's fully sent the request content, but it's also prone to race conditions even for cases where the request content is fully sent but the copy hasn't been officially marked as complete by the time libcurl tries to redirect.   ASP.NET recently began hitting this frequently on some machines in their lab runs.

While there's no perfect solution, the best solution is to go with the one that's the simplest code, the fastest code, and the deterministic outcome: the pull model.  This commit deletes a whole bunch of code that was there in support of the push model, adds just a bit of code to enable the pull model, and fixes up a couple of tests that were written to expect the push model.

Fixes https://github.com/dotnet/corefx/issues/9276
cc: @davidsh, @cipop, @ericeil, @bartonjs, @mconnew, @JunTaoLuo

@mconnew, is this going to cause any issues for WCF?
@JunTaoLuo, can you help verify that this makes the MusicStore issue go away and that it doesn't introduce any other such issues?
@davidsh, I know this isn't the model that was intended for an HttpClientHandler consuming HttpContent, but I don't see any way to do so correctly/without race conditions given the APIs available.

### CODE REVIEW

**Comment by stephentoub** [comment]
Test Innerloop Windows_NT Release Build and Test please (the leg isn't getting started)

**Comment by davidsh** [comment]
I did some research into the original design model of HttpClient to remind myself of what the actual design assumptions are. It turns out that the "pull" model for getting content to be posted (request body content) is the original design.

There is another API (protected method of HttpContent) that is used for creating content streams that are consumed/used by httpclient.

`HttpContent.CreateContentReadStreamAsync`

https://msdn.microsoft.com/en-us/library/system.net.http.httpcontent.createcontentreadstreamasync(v=vs.118).aspx

The original design of HttpClient and the handlers used that particular method (.NET Desktop does for example) as the authoritative way for handling the posting of request body data for an HTTP request.  They asked an HttpContent object to produce a stream that httpclient could read from (i.e. pull).  Then httpclient's main engine would read from that stream.

The public ReadAs* methods were designed to really be used only for consuming response data content.

When WinHttpHandler was created for CoreFx and it was a separate library/handler (still is), it was not able to use `CreateContentReadStreamAsync` and related logic that was private to http.dll  So, we ended up using the public CopyToAsync to post data from the request body.  But that ended up being a push model in some ways.

So, in summary, the "pull" model for getting data from the request body and sending it on the wire is the correct one.

**Comment by stephentoub** [comment]
@dotnet-bot test Innerloop Windows_NT Release Build and Test

**Comment by stephentoub** [comment]
Thanks, David!

**Comment by stephentoub** [comment]
@dotnet-bot test Innerloop Windows_NT Release Build and Test

**Comment by davidsh** [comment]
I noticed all the source code related to `TransportContext` and Channel Binding Token (CBT) have been removed. So, how does the TransportContext object get passed to custom HttpContent classes that want to retrieve the CBT?  Or is it that *Nix implementation doesn't provide that anyways (at least right now)?

**Comment by stephentoub** [comment]
> Or is it that *Nix implementation doesn't provide that anyways (at least right now)?

Right; the only HttpContent API I see that supports that is CopyToAsync.  Am I missing something?  There's no way I see to pass a CBT to CreateContentReadStreamAsync or ReadAsStreamAsync.

But without this change, things are significantly worse.  For example, I just tried doing a post with a megabyte of content; using the corefx echo server, and before this change it reliably fails for me every time.

**Comment by davidsh** [comment]
Hmm...I see now what is going on. I need to alter my last posting a little bit.

The original HTTP stack in Desktop is HttpWebRequest.  If you want to post request body data, you ask for the request stream (GetRequestStream*).  Then you write to it.  So, from that API perspective for the last 15+ years, it was a push model of writing data to the request body stream.

When HttpClient was created and we made an object model for HttpContent objects, we still built it on top of HttpWebRequest.  `HttpContent.CreateContentReadStreamAsync` was the model to ask a content object for a read stream to read it's data (for the purpose of posting it on the wire).  So, that part is the PULL model.

But looking at the Desktop code, that stream is read from (PULL) but then PUSHED to the request stream.  In the Desktop case, the code uses internally `CopyToAsync` and passes in the TransportContext` object.

I do see, however, that long-term, we need to reconcile this to work cross-platform. It is also more efficient as you point out in this PR to use a single model (PULL) in general.

For now, I suggest that your PR is fine in terms of simplifying the *Nix implementation. Since CBT isn't working anyways for *Nix, it shouldn't be a problem unless upper stack components are depending on the implemenation behaviors of still using CopyToAsync as a pass-thru for posting request body data.  WCF in particular has complex request body data in the form of "dynamic" streams that they create on-the-fly, and then want to "stream/send" as the request body data.

**Comment by davidsh** [comment]
For reference, here is the main logic in Desktop, System.Net.Http.HttpClientHandler:

```c#
        private void GetRequestStreamCallback(IAsyncResult ar)
        {
            RequestState state = ar.AsyncState as RequestState;
            Contract.Assert(state != null);

            try
            {
                TransportContext context = null;
                Stream requestStream = state.webRequest.EndGetRequestStream(ar, out context) as Stream;
                state.requestStream = requestStream;
                state.requestMessage.Content.CopyToAsync(requestStream, context).ContinueWithStandard(task =>
                {
                    try
                    {
                        if (task.IsFaulted)
                        {
                            HandleAsyncException(state, task.Exception.GetBaseException());
                            return;
                        }

                        if (task.IsCanceled)
                        {
                            state.tcs.TrySetCanceled();
                            return;
                        }

                        state.requestStream.Close();
                        StartGettingResponse(state);
                    }
                    catch (Exception e)
                    {
                        HandleAsyncException(state, e);
                    }

                });
            }
            catch (Exception e)
            {
                HandleAsyncException(state, e);
            }
        }
```

**Comment by mconnew** [comment]
WCF has implemented both the push and the pull mechanism as the WinRT based HttpClientHandler and WinHttpHandler each use a different mechanism so we have tested implementations for both ways. I wouldn't expect any problems for WCF.

**Comment by stephentoub** [comment]
> WCF has implemented both the push and the pull mechanism as the WinRT based HttpClientHandler and WinHttpHandler each use a different mechanism so we have tested implementations for both ways. I wouldn't expect any problems for WCF.

Excellent.  Thank you, Matt.  And I hadn't even noticed that the WinRT implementation is using ReadAsStreamAsync() as well:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs#L150
That further reduces any risk here.

**Comment by mconnew** [comment]
I've validated no regressions in basic functional tests for WCF

**Comment by stephentoub** [comment]
Thanks for the confirmation, @joelverhagen.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs [hunk 1]
```diff
@@ -1153,30 +1152,59 @@ private static ulong TransferDataFromRequestStream(IntPtr buffer, int length, Ea
...
+                        multi.EventSourceTrace("ReadAsStreamAsync completed: {0}", readStream.Status, easy: easy);
+
+                        var stateAndRequest = (Tuple<EasyRequest.SendTransferState, EasyRequest>)s;
+                        stateAndRequest.Item2._requestContentStream = readStream.GetAwaiter().GetResult();
+                        multi.EventSourceTrace("Got stream: {0}", stateAndRequest.Item2._requestContentStream.GetType(), easy: easy);
+
+                        multi.EventSourceTrace("Starting async read", easy: easy);
+                        return stateAndRequest.Item2._requestContentStream.ReadAsync(
+                           stateAndRequest.Item1._buffer, 0, Math.Min(stateAndRequest.Item1._buffer.Length, length), stateAndRequest.Item2._cancellationToken);
+                    }, Tuple.Create(sts, easy), easy._cancellationToken, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.DenyChildAttach, TaskScheduler.Default).Unwrap();
```
Comment (bartonjs) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: A) I love Tuple.Create.  One of the best uses of generic inference.

B) To me this feels like it should be chopped.. this last line is pretty verbose, though I do admit it's boilerplate.

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: A) :smile:
B) Thanks.  If I have other reasons to resubmit it, I'll reformat the line.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs [hunk 2]
```diff
@@ -1207,26 +1235,47 @@ private static Interop.Http.CurlSeekResult CurlSeekCallback(IntPtr context, long
...
+                        // If we don't have a stream yet, we can't seek.
+                        if (easy._requestContentStream == null)
                         {
-                            // Dump any state associated with the old stream's position
-                            easy._sendTransferState?.SetTaskOffsetCount(null, 0, 0);
-
-                            // Restart the transfer
-                            easy._requestContentStream.Run();
+                            CurlHandler.EventSourceTrace("No request stream exists yet. Can't seek", easy: easy);
+                            return Interop.Http.CurlSeekResult.CURL_SEEKFUNC_CANTSEEK;
```
Comment (bartonjs) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Later if curl asks you to seek to 0 you null the stream.  Is checking for seek-to-zero in the already null case interesting?  Like... if they called seek twice to the same value?

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs [hunk 3]
```diff
@@ -1207,26 +1235,47 @@ private static Interop.Http.CurlSeekResult CurlSeekCallback(IntPtr context, long
...
+                            CurlHandler.EventSourceTrace("Seeking on the existing stream", easy: easy);
+                            easy._requestContentStream.Seek(offset, (SeekOrigin)origin);
                             return Interop.Http.CurlSeekResult.CURL_SEEKFUNC_OK;
                         }
-                        else
+
+                        // The stream isn't seekable.  Now we start getting into shakier ground.
+                        // Most of the time the seek callback is used, it's because libcurl is rewinding
+                        // to the beginning of the stream due to a redirect, an auth challenge, etc. (other
+                        // cases where it might try to seek elsewhere would be, e.g., with a Range header).
```
Comment (bartonjs) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Re: Range

if (offset == 0 && origin == SeekOrigin.Current) { nop }

if (offset > 0 && origin == SeekOrigin.Current) { can't we just call Read() to move the stream to the right position? Or will curl do that for us? }

Both of those assume that curl's origin has a "Current" mode.  Or that we know we haven't read anything yet, so that Begin == Current.

Maybe neither is worth addressing; but they seem like they'd be straightforward to handle (if we can detect the state)

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: At least according to the docs, libcurl currently only uses SeekOrigin.Begin (aka SEEK_SET):
https://curl.haxx.se/libcurl/c/CURLOPT_SEEKFUNCTION.html
```
The function shall work like fseek(3) or lseek(3) and it gets SEEK_SET, SEEK_CUR or SEEK_END as argument for origin, although libcurl currently only passes SEEK_SET
```

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs [hunk 4]
```diff
@@ -199,6 +199,7 @@ public static IEnumerable<object[]> UseCallback_ValidCertificate_ExpectedValuesD
             }
         }
 
+        [PlatformSpecific(PlatformID.Windows)] // CopyToAsync(Stream, TransportContext) isn't used on unix
```
Comment (bartonjs) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs]: "isn't used", or "throws?"  And is it a CopyToStream problem, or a ChannelBinding problem?

Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs]: "isn't used".  We call ReadAsStreamAsync() (like the WinRT implementation) rather than calling CopyToAsync(Stream, TransportContext) (like the WinHttp implementation)

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 5]
```diff
@@ -1119,6 +1119,34 @@ public static IEnumerable<object[]> VerifyUploadServersStreamsAndExpectedData
...
 
+        [OuterLoop] // takes several seconds
+        [Theory]
+        [InlineData(302, false)]
+        [InlineData(307, true)]
+        public async Task PostAsync_Redirect_LargePayload(int statusCode, bool expectRedirectToPost)
+        {
+            using (var fs = new FileStream(Path.Combine(Path.GetTempPath(), Path.GetTempFileName()), FileMode.Create, FileAccess.ReadWrite, FileShare.None, 0x1000, FileOptions.DeleteOnClose))
+            {
+                string contentString = string.Join("", Enumerable.Repeat("Content", 100000));
```
Comment (bartonjs) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: 700k characters?  Do computers even have that much RAM? :smile:

I'm a little surprised that you didn't just call fs.Write 100000 times with the UTF-32 string "Content".

Not really important for a test, I suppose.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs [hunk 6]
```diff
@@ -1207,26 +1235,47 @@ private static Interop.Http.CurlSeekResult CurlSeekCallback(IntPtr context, long
...
+                            return Interop.Http.CurlSeekResult.CURL_SEEKFUNC_CANTSEEK;
+                        }
 
-                            CurlHandler.EventSourceTrace("Seek successful", easy: easy);
+                        // If the stream is seekable, which is a very common case, everyone is happy.
+                        // Simply seek on the stream.
+                        if (easy._requestContentStream.CanSeek)
+                        {
+                            CurlHandler.EventSourceTrace("Seeking on the existing stream", easy: easy);
+                            easy._requestContentStream.Seek(offset, (SeekOrigin)origin);
```
Comment (bartonjs) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: I'm waffling on whether it's a dumb question to ask, so I'll ask it anyways: I know you've quoted the docs as saying that it'll only be 0 (or 1 or 2 if they get ambitious); is it worth asserting here that SeekOrigin.HasValue(origin)?

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs (hunk 3 above)
Comment (bartonjs) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Sounds convincing.  Do we happen to have any tests which use a non-seekable stream and a HTTP Range GET?  So as to see if this is something curl shrugs off and does the read-to-the-bit-bucket on (or if it might we worth special-casing the "we got a call to seek before any call to read")...

I'll also happily accept "eh, it's so niche that I don't want to risk destabilizing something for it right now, but it's a reasonable thing to consider in the future" or other actual (or dismissive) deferral.

**Comment by bartonjs** [comment]
I'm happy with this as-is.  I think all of my questions can be pushed off as future enhancement, or are just comments from me just being a chatterbox tonight (e.g. comparing to 640k of RAM).

LGTM.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: I don't think we have a test for it, but I looked to see how libcurl handles it if it gets back CURL_SEEKFUNC_CANTSEEK, and it just reads-and-discards until the relevant position, so it's a perf optimization rather than a behavioral flaw.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Hmm, interesting point.  I don't think it'll do that.  And at this point with most folks signed off, I'd rather not add such additional logic.  We can certainly add that easily if we ever see it happening.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs (hunk 6 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Interesting thought, but I don't think it's worth asserting.  We just pass through to the stream whatever we're given.  If it's invalid, it may throw, in which case the request will fail as it should, and if it doesn't throw, then who are we to say it's wrong :smile:

**Comment by stephentoub** [comment]
Excellent.  Thanks for the solid review, Jeremy.

**Comment by JunTaoLuo** [comment]
I have confirmed that this fixes the race condition seen in our tests. With the custom build provided by @stephentoub, our test suite is now passing consistently. Thanks!

**Comment by stephentoub** [comment]
Thanks, @JunTaoLuo!
